File: Sources/Commands/Add.swift
//
//  Alias.swift
//  
//
//  Created by Owen Cruz-Abrams on 9/21/23.
//

import Foundation
import ArgumentParser
import Darwin

struct AddOptions: ParsableArguments {
    @Argument(help: "The alias to be added. This is the command that will be given in the future, not the longform version being aliased to.")
    var alias: String
    
    @Argument(help: "The long-form command to be aliased to `alias`.")
    var command: String?
    
    @Option(help: "Description/title of the alias, to be put on the preceding line in the file.")
    var description: String?
    
    @Flag(name: [.long], help: "Reload the terminal (a.k.a run \u{001b}[1msource\u{001b}[0m) after adding the alias.")
    var reload = false
    
    @Flag(name: [.long], help: "Produce no output.")
    var noOutput = false
    
    
}

extension Ally {
    
    struct Add: ParsableCommand {
        func conditionalPrint(_ str: String) {
            if !options.noOutput {
                print(str)
            }
        }
        static var configuration = CommandConfiguration(abstract: "Add an alias to the zsh config.")
        
        @OptionGroup var options: AddOptions
        
        mutating func run() throws {
            if options.alias.localizedCaseInsensitiveContains("=") {
                let newOpts = options.alias.split(separator: "=", maxSplits: 1).map{String($0)}
                options.alias = newOpts[0]
                options.command = newOpts[1]
            }
            if let command = options.command {
                // First, check to ensure that .ally exists. If not, run ally init --no-output
                let dotFileLocation = Ally.dotFileLocation
                if FileManager.default.fileExists(atPath: dotFileLocation.relativePath) {
                    // Continue
                } else {
                    let yesorno = input("[SYSTEM] Ally has not yet been installed, would you like to install it? (y/n)")
                    if yesorno != "y" {
                        Ally.Add.exit(withError: nil)
                    }
                    do {
                        try safeShell("ally init --no-output")
                        conditionalPrint("[SYSTEM] Ally has been installed and initialized.")
                    } catch {
                        conditionalPrint("There was an error installing ally: \(error.localizedDescription). Exiting now.")
                        Ally.Add.exit(withError: nil)
                    }
                }
                // Ally is now installed, so we can procede.
                // Step #1: Create the new alias line
                var aliasLine: String
                if let description = options.description {
                    aliasLine = """
    
    # \(description)
    alias \(options.alias)="\(command)"
    """
                } else {
                    aliasLine = """
    
    alias \(options.alias)="\(command)"
    """
                }
                
                // Step #2: Add it to .ally
                do {
                    var fileContents = try String(contentsOf: dotFileLocation, encoding: .utf8)
                    // Check to see if the alias itself is already in there
                    if fileContents.range(of: "alias \(options.alias)") == nil {
                        fileContents += aliasLine
                        try fileContents.write(to: dotFileLocation, atomically: true, encoding: .utf8)
                    } else {
                        conditionalPrint("[WARNING] alias \u{001b}[1m\(options.alias)\u{001b}[0m ALREADY EXISTS. Skipping.")
                    }
                    
                } catch {
                    conditionalPrint("[ERROR] Saving alias to disk: \(error.localizedDescription)")
                }
                
                if options.reload {
                    //                system("source $HOME/.ally")
                    // TODO: Reload shell
                    conditionalPrint("Shell reloaded.")
                } else {
                    conditionalPrint("[NOTE]: Shell may require reload.")
                }
            }
            else {
                print("[ERROR] No value provided to alias '\u{001b}[1m\(options.alias)\u{001b}[0m' TO. Please provide a long-form command to set the alias to.")
            }
            
        }
    }
    
}

File: src/commands/add.py
import argparse
import os
import subprocess

class AddOptions:
    def __init__(self):
        parser = argparse.ArgumentParser(description="Add an alias to the zsh config.")
        parser.add_argument("alias", help="The alias to be added. This is the command that will be given in the future, not the longform version being aliased to.")
        parser.add_argument("command", nargs='?', help="The long-form command to be aliased to `alias`.")
        parser.add_argument("--description", help="Description/title of the alias, to be put on the preceding line in the file.")
        parser.add_argument("--reload", action="store_true", help="Reload the terminal (a.k.a run \u001b[1msource\u001b[0m) after adding the alias.")
        parser.add_argument("--no-output", action="store_true", help="Produce no output.")
        self.args = parser.parse_args()

class AllyAdd:
    def __init__(self):
        self.options = AddOptions().args

    def conditional_print(self, message):
        if not self.options.no_output:
            print(message)

    def run(self):
        if "=" in self.options.alias:
            new_opts = self.options.alias.split("=", 1)
            self.options.alias = new_opts[0]
            self.options.command = new_opts[1]

        if self.options.command:
            dot_file_location = os.path.expanduser("~/.ally")
            if not os.path.exists(dot_file_location):
                yes_or_no = input("[SYSTEM] Ally has not yet been installed, would you like to install it? (y/n)")
                if yes_or_no != "y":
                    return
                try:
                    subprocess.run(["ally", "init", "--no-output"], check=True)
                    self.conditional_print("[SYSTEM] Ally has been installed and initialized.")
                except subprocess.CalledProcessError as e:
                    self.conditional_print(f"There was an error installing ally: {e}. Exiting now.")
                    return

            alias_line = f'\nalias {self.options.alias}="{self.options.command}"\n'
            if self.options.description:
                alias_line = f'\n# {self.options.description}\n{alias_line}'

            try:
                with open(dot_file_location, "a") as file:
                    file_contents = file.read()
                    if f"alias {self.options.alias}=" not in file_contents:
                        file.write(alias_line)
                    else:
                        self.conditional_print(f"[WARNING] alias \u001b[1m{self.options.alias}\u001b[0m ALREADY EXISTS. Skipping.")
            except Exception as e:
                self.conditional_print(f"[ERROR] Saving alias to disk: {e}")

            if self.options.reload:
                self.conditional_print("Shell reloaded.")
            else:
                self.conditional_print("[NOTE]: Shell may require reload.")
        else:
            print(f"[ERROR] No value provided to alias '\u001b[1m{self.options.alias}\u001b[0m' TO. Please provide a long-form command to set the alias to.")

if __name__ == "__main__":
    AllyAdd().run()
